`data_defined_status_domain` is incompatible with `llfio` on GCC

**Describe the bug**
GCC (incorrectly?) assesses an atomic constraint do depend on itself within the `std::is_constructible` type trait while trying to evaluate whether `outcome::result` has enabled implicit constructors or not. The recursion path is roughly `is_constructible -> status_code -> quick_status_code_from_enum -> status_enum -> status_enum_definition -> default_initializable -> is_constructible`.

**To Reproduce / MCVE**
```
#include <outcome/experimental/status_result.hpp>
#include <outcome/try.hpp>
#include <status-code/system_code.hpp>

#include <dplx/cncr/data_defined_status_domain.hpp>
//
#include <llfio/llfio.hpp>
```
`g++ -DLLFIO_EXPERIMENTAL_STATUS_CODE=1 -DLLFIO_HEADERS_ONLY=0 -DOUTCOME_USE_SYSTEM_STATUS_CODE=1 -DQUICKCPPLIB_ENABLE_VALGRIND=1 -fconcepts-diagnostics-depth=0 -ftemplate-backtrace-limit=0 -fno-omit-frame-pointer -g -std=c++20 -I"x64-linux-gcc/vcpkg_installed/x64-linux/include" -c "mcve.cpp"`

**Expected behavior**
The code should compile.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment**
 - OS: Ubuntu LTS 22.04
 - compiler: gcc 11.3
 - Project Version: 0.0.0-alpha.7

**Additional context**
```
/usr/include/c++/11/type_traits: In substitution of ‘template<class T>  requires  outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> constexpr outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::basic_result<T>(T&&, outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::error_converting_constructor_tag) [with T = quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>]’:
/usr/include/c++/11/type_traits:946:30:   required from ‘struct std::__is_constructible_impl<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/type_traits:952:12:   required from ‘struct std::is_constructible<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/type_traits:3173:38:   required from ‘constexpr const bool std::is_constructible_v<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/concepts:137:30:   required by substitution of ‘template<class Enum>  requires (default_initializable<Enum>) && (movable<Enum>) && (equality_comparable<Enum>) struct dplx::cncr::status_enum_definition [with Enum = llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state]’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/status-code/errored_status_code.hpp:263:3:   [ skipping 3 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/usr/include/c++/11/type_traits:1457:12:   required from ‘struct std::is_convertible<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > > >’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/config.hpp:252:62:   required from ‘constexpr const bool outcome_v2::detail::_is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>::value’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/config.hpp:262:119:   required from ‘constexpr const bool outcome_v2::detail::is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:77:73:   required from ‘constexpr const bool outcome_v2::detail::result_predicates<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > > >::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:310:23:   required from ‘constexpr const bool outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::predicate::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:416:3:   required by substitution of ‘template<class T>  requires  outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> constexpr outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::basic_result<T>(T&&, outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::error_converting_constructor_tag) [with T = quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>]’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/llfio/v2.0/byte_io_multiplexer.hpp:900:78:   required from here
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:417:13:   required by the constraints of ‘template<class R, class S, class NoValuePolicy> template<class T>  requires  outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> constexpr outcome_v2::basic_result<R, S, NoValuePolicy>::basic_result(T&&, outcome_v2::basic_result<R, S, NoValuePolicy>::error_converting_constructor_tag)’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:416:21: error: satisfaction of atomic constraint ‘outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> [with T = T; R = R; S = S; NoValuePolicy = NoValuePolicy]’ depends on itself
  416 |   OUTCOME_TREQUIRES(OUTCOME_TPRED(predicate::template enable_error_converting_constructor<T>))
      |                     ^
In file included from /home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/experimental/status_result.hpp:28,
                 from /home/me/.vs/concrete/src/dplx/cncr/data_defined_status_domain.test.cpp:2:
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp: In instantiation of ‘constexpr const bool outcome_v2::detail::result_predicates<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > > >::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’:
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:310:23:   required from ‘constexpr const bool outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::predicate::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:416:3:   required by substitution of ‘template<class T>  requires  outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> constexpr outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::basic_result<T>(T&&, outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::error_converting_constructor_tag) [with T = quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::const_buffer_type>]’
/usr/include/c++/11/type_traits:946:30:   required from ‘struct std::__is_constructible_impl<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/type_traits:952:12:   required from ‘struct std::is_constructible<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/type_traits:3173:38:   required from ‘constexpr const bool std::is_constructible_v<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/usr/include/c++/11/concepts:137:30:   [ skipping 5 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/usr/include/c++/11/type_traits:1457:12:   required from ‘struct std::is_convertible<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > > >’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/config.hpp:252:62:   required from ‘constexpr const bool outcome_v2::detail::_is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>::value’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/config.hpp:262:119:   required from ‘constexpr const bool outcome_v2::detail::is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:77:73:   required from ‘constexpr const bool outcome_v2::detail::result_predicates<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > > >::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:310:23:   required from ‘constexpr const bool outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::predicate::enable_error_converting_constructor<llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:416:3:   required by substitution of ‘template<class T>  requires  outcome_v2::basic_result<R, S, NoValuePolicy>::predicate::enable_error_converting_constructor<T> constexpr outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::basic_result<T>(T&&, outcome_v2::basic_result<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, outcome_v2::experimental::policy::status_code_throw<quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>, system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, void> >::error_converting_constructor_tag) [with T = quickcpplib::_776de923::span::span<llfio_v2_85eb6946::byte_io_multiplexer::buffer_type>]’
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/llfio/v2.0/byte_io_multiplexer.hpp:900:78:   required from here
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:77:73: error: the value of ‘outcome_v2::detail::is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’ is not usable in a constant expression
   77 |     && ((!detail::is_implicitly_constructible<value_type, T> && detail::is_implicitly_constructible<error_type, T>)  // is unambiguously for error type
      |                                                                 ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/basic_result.hpp:28,
                 from /home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/experimental/status_result.hpp:28,
                 from /home/me/.vs/concrete/src/dplx/cncr/data_defined_status_domain.test.cpp:2:
/home/me/.vs/concrete/build/x64-linux-gcc/vcpkg_installed/x64-linux/include/outcome/config.hpp:262:53: note: ‘outcome_v2::detail::is_implicitly_constructible<system_error2::errored_status_code<system_error2::erased<llfio_v2_85eb6946::detail::file_io_error_value_type<long int> > >, llfio_v2_85eb6946::byte_io_multiplexer::_unsynchronised_io_operation_state>’ used in its own initializer
  262 |   template <class T, class U> static constexpr bool is_implicitly_constructible = _is_implicitly_constructible<T, U>::value;
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/dplx/cncr/data_defined_status_domain.hpp
+++ b/src/dplx/cncr/data_defined_status_domain.hpp
@@ -53,14 +53,20 @@ struct status_enum_definition_defaults
 };
 
 template <typename Enum>
-    requires std::default_initializable<Enum> && std::movable<
-            Enum> && std::equality_comparable<Enum>
-struct status_enum_definition;
+struct status_enum_definition
+{
+};
 
 // clang-format off
 template <typename T>
 concept status_enum
-        = !system_error::is_status_code<T>::value
+        = requires {
+            typename status_enum_definition<T>::code;
+            requires std::same_as<T, typename status_enum_definition<T>::code>;
+            requires std::default_initializable<T>;
+            requires std::movable<T>;
+            requires std::equality_comparable<T>;
+        }
         && std::constructible_from<
                 std::span<status_enum_value_descriptor<T> const>,
                 decltype(status_enum_definition<T>::values)>

